### ISSUE TITLE
Start work on the transition to python3

### ISSUE THREAD

**Issue by santoshphilip** [santoshphilip/eppy#39]

**Comment by santoshphilip** [comment]
fully transitioned  to python3. This was done in branch "develop_lint_3k".
In this branch the following was done:
- ran pylint on all python files
- transitioned to python3 using 2to3 program
- successfully ran unit tests
- one test is failing due to a floating point mismatch.
- also tested all the documentation scripts
- third party package pydot is not working in python3

Things to do:
- clean up and make ready for release
- See if you can make pydot work in python3
- announce once it is done

**Comment by Jerome-github** [comment]
Hi.

Any news about Python3 support ?

AFAIU, there is a pydot version that works with Python3: [pydot3k](https://pypi.python.org/pypi/pydot3k/1.0.17). I can find neither the code nor the releases in github, but it appears on pypi. [pydotplus](https://pypi.python.org/pypi/pydotplus/2.0.2) might be an alternative.

Thanks.

**Comment by santoshphilip** [comment]
Thanks for the information about pydot. I will check it out.
Do you think you can help with the python3 ?
I have not actually used it, but I believe it is fully functional except for the pydot stuff.

Let me know if you have the time to look into it, and I can tell you how to start.

**Comment by santoshphilip** [comment]
need to get this ready for release

**Comment by Jerome-github** [comment]
Well, we have a R&D project that uses E+ in batch mode. We're thinking of using eppy for the idf file management. Since the development is in an early phase, we'd like to use Python3 and I think eppy is the only lib we need that is not available yet, so before we start, I'd like to know if there's any chance it will be released in the next months.

Of course, I'll be happy to help if I can, but I don't really see how. 

If you tell us the few things we need to know to install the Python 3 version of eppy, maybe we can beta test it. AFAIU, we'd need to checkout branch develop_lint_3k, maybe change "pydot>1.0" into "pydot3k>1.0". Anything else ? Note that we may be doing this in a few weeks if not month, so don't expect quick feedback. Besides, our use may be much more limited than automated test coverage.

Apparently, the unit tests ran smoothly, so there's ground to be confident. Is there anything else to do than beta-testing ? I mean you did the porting already. And I have scarce knowledge about packaging. I don't even know what pydot is for. Is it optional ?

**Comment by santoshphilip** [comment]
- pydot is optional. Needed only if you want to draw the HVAC diagrams
- to use the python3 version:
    - git checkout develop_lint_3k
    - goto folder p3
    - everything within p3/eppy is the python3 version of eppy
    - follow the documentation of the python2 version, use python3 syntax where needed
- it is not ready for a "pip install". Let me know if you need to do so.

If you are planning on using the python3 version, I would be very motivated to complete it. Right now it is dragging since I don't use python3 and did not hear from anyone needing it.

I am not very clear about the housekeeping needed to release both python2 and python3 version. I'll research how other projects are doing it.

**Comment by Jerome-github** [comment]
Hi.

We peaked into the .ipynb files, and an intern here gave eppy a go for another project. It looks really promising. It should save us time and headaches. All we wanted to do in about three lines of code. Congratulations. We're definitely going for it.

As developers, we may not need eppy as "pip install", but ultimately we'll need it to distribute our software. I know it can take time and it's not the funniest part of the dev. This is why I'm asking in advance, so that we choose the right python version to develop.

> If you are planning on using the python3 version, I would be very motivated to complete it. Right now it is dragging since I don't use python3 and did not hear from anyone needing it.

Great! Thanks.

> I am not very clear about the housekeeping needed to release both python2 and python3 version. I'll research how other projects are doing it.

Python 3 code may be easier and cleaner if dropping Python 2 support, but then you'd have to backport fixes to Python 2 until projects based on Python 2 join the move. Alternatively, you may be able to get away with only [one branch supporting both Python versions](https://docs.python.org/3/howto/pyporting.html).

If my understanding is correct, you converted all files to Python 3 in a subdirectory. (You could have converted them in place, since you use git, and you would have been able to see the differences the easy way.)

The page I linked to suggests upgrading the code to remove Python 2 specificities and make it Python 3 compliant. This would introduce good practices into the Python 2 codebase in the process.

Depending on the foreseen lifetime of the Python 2 codebase of eppy, the complexity of the code and the frequency of updates/modifications, you may prefer to maintain two branches and backport changes from time to time rather than going through all you code to ensure compatibility with both Python 2 and 3.

I don't think I'd participate directly in this, but whichever option you choose, feel free to leave a note here if you're stuck on something. I'll be happy to help if I'm able to.

Thanks.

**Comment by santoshphilip** [comment]
Thanks for the comments.
I have been reading up on this and I will have to make a decision of which path way to take. 
I'll also contact some of the existing users to see how badly they need to stick to python2.

It should  be possible to to have a python3 release over the next couple of weeks.

**Comment by JasonGlazer** [comment]
I am in the middle of using Python 2.7.5 and would prefer that Eppy stayed compatible with Python 2 for the foreseeable future. It would be fine if it was also compatible with Python 3, that may be were the industry is going (slowly) but it is important that if bugs are found that they are fixed in the Python 2 version of Eppy for the next year or so.

**Comment by santoshphilip** [comment]
for the foreseeable future, let us continue with python2 development and use 2to3 to generate the python3 code.
pylint will be used to ensure that python2 development stays amenable for python3 conversion

At some point thought will have to be put into transitioning completely to python3. 
Having the code developed in python3 is not mission-critical, since eppy does not need any of the python3 features yet.

This issue will be closed once the python3 version is released

**Comment by santoshphilip** [comment]
pydot3k works in the place of pydot. 
So now all of the code works in python3

**Comment by santoshphilip** [comment]
made a branch **i39_python3** (from develop) to explore how the python3 version should be :

- generated
- tested 
- released. 

looks like setuptools has everything needed to do this.

**Comment by santoshphilip** [comment]
- At this point, the p3 folder has a fully functioning eppy in python3
- All future releases of eppy will work in python3
- major tasks raised in this issue have been completed sucessfully.
- this issue can be closed

Open a new issue for the following:
- We need a better process flow to generate the python3 files
- We also need pip to install python3 when appropriate

**Comment by Jerome-github** [comment]
Thanks !

Here's the issue for pip install: https://github.com/santoshphilip/eppy/issues/64

I let you proceed with the "process flow" issue, as I don't know what is done and what is missing / desirable. (Or maybe you'd like to merge both issues into one?)
